### PR TITLE
 [ELY-1587] ASN.1 decoding extended - UTF8String, BMPString...

### PR DESCRIPTION
**Pull request by hkalina**
https://issues.jboss.org/browse/ELY-1587
(depends on https://issues.jboss.org/browse/ELY-1572 - included)

### CODE REVIEW

File: src/main/java/org/wildfly/security/asn1/ASN1Decoder.java [hunk 1]
```diff
@@ -172,21 +172,81 @@
...
     byte[] decodePrintableStringAsBytes() throws ASN1Exception;
 
+    /**
+     * Decode the next ASN.1 element as an UniversalString.
+     *
+     * @return the decoded printable string as a string
+     * @throws ASN1Exception if the next element is not a printable string
+     */
+    default String decodeUniversalString() throws ASN1Exception {
+        throw new UnsupportedOperationException();
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/asn1/ASN1Decoder.java]: It should be documented that the implementation throws this exception.  Something like this:

```java
@throws UnsupportedOperationException if this implementation does not provide this method
```

File: src/main/java/org/wildfly/security/asn1/ASN1Decoder.java [hunk 2]
```diff
@@ -172,21 +172,87 @@
...
      *
      * @return the decoded printable string as a byte array
      * @throws ASN1Exception if the next element is not a printable string
      */
     byte[] decodePrintableStringAsBytes() throws ASN1Exception;
 
+    /**
+     * Decode the next ASN.1 element as an UniversalString.
+     *
+     * @return the decoded printable string as a string
```
[the remark this answers is not in the document]
Comment (hkalina) [reply src/main/java/org/wildfly/security/asn1/ASN1Decoder.java]: fixed

**Comment by mchoma** [comment]
Ack - not a RFE. 

Why was it necessary to turn off source compatibility check?

**Comment by hkalina** [comment]
@mchoma because of adding new (default) method into interface - even through japicmp should ignore METHOD_NEW_DEFAULT, because of bug it is also reported as source-incompatible METHOD_ADDED_TO_INTERFACE (see https://github.com/siom79/japicmp/issues/201 ) - check can be re-enabled after new version of japicmp release - it is fixed in upstream.

**Comment by darranl** [comment]
@fjuma did you want to do the second review on this one?  Looks fine to me but you have worked quite a bit in this area.

**Pull request merged**